cloverage/src/cloverage/coverage.clj

```diff
@@ -54,23 +55,19 @@
          (finally
            (copy :leiningen/skipped-test :test)))))
 
-(defmacro with-coverage [libs & body]
-  `(binding [*covered* (atom [])]
-     (println "Capturing code coverage for" ~libs)
-     (doseq [lib# ~libs]
-       (instrument #'track-coverage lib#))
-     ~@body
-     (gather-stats @*covered*)))
+(defn covered []
+  (mapv (fn [{:keys [^AtomicInteger hits] :as form}]
+          (merge form {:hits (.get hits) :covered (pos? (.get hits))}))
+        @*covered*))
 
 (defn cover
   "Mark the given file and line in as having been covered."
   [idx]
-  (let [covered (swap! *covered* #(if-let [{:keys [hits] :as data} (nth % idx nil)]
-                                    (assoc % idx (assoc data :covered true :hits (inc (or hits 0))))
-                                    %))]
-    (when-not (nth covered idx nil)
-      (log/warn (str "Couldn't track coverage for form with index " idx
-                     " covered has " (count covered) ".")))))
+  ;; Note well that this function is written to have minimal overhead. Make sure
+  ;; there are no reflection warnings and especially beware introducing any
+  ;; unnecessary coordination here – it can greatly affect performance when
+  ;; instrumenting tests that use multithreading.
+  (.getAndIncrement ^AtomicInteger (get (nth (.deref ^IDeref *covered*) idx) :hits)))
```

Comment:
Huh -- `@` introduces reflection? I guess it would since there are multiple things it can do; I never considered that.

Reply:
It doesn't, but it does expand to a call to `clojure.core/deref` – instead I'm just calling the `.deref` interface method. It's a micro-optimisation to avoid the `instance?` check in `deref`.